### PR TITLE
Produce RunLengthEncodedBlock in RowBlockBuilder when all values are null

### PR DESCRIPTION
**Pull request by lukasz-stec**
## Description

Add support in RowBlockBuilder for producing RunLengthEncodedBlock if all positions are null.

> Is this change a fix, improvement, new feature, refactoring, or other?

improvement
> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)

changes only in RowBlockBuilder, AbstractRowBlock classes in the trino-spi, not affecting API.

> How would you describe this change to a non-technical end user or system administrator?

## Related issues, pull requests, and links

Similar to https://github.com/trinodb/trino/pull/12043

## Documentation

( ) No documentation is needed.
( ) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

( ) No release notes entries required.
( ) Release notes entries required with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

File: core/trino-spi/src/main/java/io/trino/spi/block/AbstractRowBlock.java [hunk 1]
```diff
@@ -72,7 +72,7 @@ public String getEncodingName()
     }
 
     @Override
-    public final Block copyPositions(int[] positions, int offset, int length)
```
Comment (sopel39) [inline core/trino-spi/src/main/java/io/trino/spi/block/AbstractRowBlock.java:75]: nit: separate commit

File: core/trino-spi/src/main/java/io/trino/spi/block/RowBlockBuilder.java [hunk 2]
```diff
@@ -224,6 +228,11 @@ public Block build()
         for (int i = 0; i < numFields; i++) {
             fieldBlocks[i] = fieldBlockBuilders[i].build();
         }
+
+        if (!hasNonNullRow) {
```
Comment (sopel39) [inline core/trino-spi/src/main/java/io/trino/spi/block/RowBlockBuilder.java:232]: move this before:
```
        Block[] fieldBlocks = new Block[numFields];
        for (int i = 0; i < numFields; i++) {
            fieldBlocks[i] = fieldBlockBuilders[i].build();
        }
```

and just call:
```
        if (!hasNonNullRow) {
            return nullRle(positionCount);
        }
```

**Review by sopel39 — APPROVED** [review]
% comments

File: core/trino-spi/src/test/java/io/trino/spi/block/TestRowBlockBuilder.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        // multiple nulls
+        assertIsNullRle(blockBuilder().appendNull().appendNull().build(), 2);
+    }
+
+    @Test
+    public void testCopyNullRle()
+    {
+        BlockBuilder blockBuilder = blockBuilder().appendNull().appendNull();
+        assertIsNullRle(blockBuilder.copyPositions(new int[] {0}, 0, 1), 1);
+        assertIsNullRle(blockBuilder.getRegion(0, 1), 1);
```
[the remark this answers is not in the document]
Comment (lukasz-stec) [reply core/trino-spi/src/test/java/io/trino/spi/block/TestRowBlockBuilder.java:39]: merged two methods into `testBuilderProducesNullRleForNullRows`

File: core/trino-spi/src/test/java/io/trino/spi/block/TestRowBlockBuilder.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+import org.testng.annotations.Test;
+
+import static io.trino.spi.type.BigintType.BIGINT;
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertTrue;
+
+public class TestRowBlockBuilder
+{
+    @Test
+    public void testBuildNullRle()
```
[the remark this answers is not in the document]
Comment (lukasz-stec) [reply core/trino-spi/src/test/java/io/trino/spi/block/TestRowBlockBuilder.java]: renamed to `testBuilderProducesNullRleForNullRows`

File: core/trino-spi/src/main/java/io/trino/spi/block/RowBlockBuilder.java (hunk 2 above)
Comment (lukasz-stec) [reply core/trino-spi/src/main/java/io/trino/spi/block/RowBlockBuilder.java]: done

File: core/trino-spi/src/main/java/io/trino/spi/block/AbstractRowBlock.java (hunk 1 above)
Comment (lukasz-stec) [reply core/trino-spi/src/main/java/io/trino/spi/block/AbstractRowBlock.java:75]: IMO it's easier to see the reason for this change if it's part of the commit that overrides this method.

**Review by lukasz-stec — COMMENTED** [review]
comments addressed

**Comment by sopel39** [comment]
@lukasz-stec there are test failures

**Comment by lukasz-stec** [comment]
> @lukasz-stec there are test failures

`io.trino.spi.TestSpiBackwardCompatibility#testSpiSingleVersionBackwardCompatibility` fails.
Looks like it doesn't support `final`  keyword removal which is a backward compatible change as far as I can tell.
@kokosing is this intended behavior?

File: core/trino-spi/src/main/java/io/trino/spi/block/AbstractRowBlock.java (hunk 1 above)
Comment (sopel39) [reply core/trino-spi/src/main/java/io/trino/spi/block/AbstractRowBlock.java:75]: I would remove this change for now as it causes test failure and create an issue in Github that SPI checker shouldn't fail for such case

**Comment by kokosing** [comment]
> Looks like it doesn't support final keyword removal which is a backward compatible change as far as I can tell.
@kokosing is this intended behavior?

It was not intended.

File: core/trino-spi/src/main/java/io/trino/spi/block/RowBlockBuilder.java [hunk 5]
```diff
@@ -220,10 +224,14 @@ public Block build()
...
             throw new IllegalStateException("Current entry must be closed before the block can be built");
         }
+        if (!hasNonNullRow) {
+            return nullRle(positionCount);
+        }
         Block[] fieldBlocks = new Block[numFields];
         for (int i = 0; i < numFields; i++) {
             fieldBlocks[i] = fieldBlockBuilders[i].build();
         }
+
```
Comment (sopel39) [inline core/trino-spi/src/main/java/io/trino/spi/block/RowBlockBuilder.java:234]: undo, unrelated change

**Review by sopel39 — APPROVED** [review]
% comments

**Review by lukasz-stec — COMMENTED** [review]
`copyPositions` removed due to SPI compatibility check failing

**Comment by lukasz-stec** [comment]
since https://github.com/trinodb/trino/pull/12062 is merged i rebased on the master and re-added `RowBlockBuilder:copyPositions`

**Review by sopel39 — APPROVED** [review]
